### PR TITLE
CCA: Add support for AES CIPHER keys

### PR DESCRIPTION
**Pull request by ifranzki**

### CODE REVIEW

File: usr/lib/cca_stdll/cca_specific.c [hunk 1]
```diff
@@ -3388,70 +3421,76 @@ CK_RV token_specific_aes_xts_key_gen(STDLL_TokData_t *tokdata, TEMPLATE *tmpl,
...
+    *is_opaque = TRUE;
+
+    if (cca_data->aes_key_mode == AES_KEY_MODE_CIPHER)
+        rc = cca_build_aes_cipher_token(tokdata, tmpl,
+                                        key_token, &key_token_len);
+    else
+        rc = cca_build_aes_data_token(tokdata, key_size / 2,
+                                      key_token, &key_token_len);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Faild to build CCA key token");
```
Comment (jschmidb) [inline usr/lib/cca_stdll/cca_specific.c:3512]: Failed

File: usr/lib/cca_stdll/cca_specific.c [hunk 2]
```diff
@@ -3388,70 +3421,76 @@ CK_RV token_specific_aes_xts_key_gen(STDLL_TokData_t *tokdata, TEMPLATE *tmpl,
...
+        /* Caller will free retuned aes_key */
```
Comment (jschmidb) [inline usr/lib/cca_stdll/cca_specific.c:3513]: returned

File: usr/lib/cca_stdll/cca_specific.c [hunk 3]
```diff
@@ -5888,80 +6038,207 @@ CK_RV token_specific_rsa_pss_verify(STDLL_TokData_t *tokdata,
...
+    *is_opaque = TRUE;
+
+    if (cca_data->aes_key_mode == AES_KEY_MODE_CIPHER)
+        rc = cca_build_aes_cipher_token(tokdata, tmpl,
+                                        *aes_key, &key_token_len);
+    else
+        rc = cca_build_aes_data_token(tokdata, key_size,
+                                      *aes_key, &key_token_len);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Faild to build CCA key token");
```
Comment (jschmidb) [inline usr/lib/cca_stdll/cca_specific.c:6562]: Failed

File: usr/lib/cca_stdll/cca_specific.c (hunk 1 above)
Comment (ifranzki) [reply usr/lib/cca_stdll/cca_specific.c:3512]: fixed

File: usr/lib/cca_stdll/cca_specific.c (hunk 2 above)
Comment (ifranzki) [reply usr/lib/cca_stdll/cca_specific.c:3513]: fixed

File: usr/lib/cca_stdll/cca_specific.c (hunk 3 above)
Comment (ifranzki) [reply usr/lib/cca_stdll/cca_specific.c:6562]: fixed

File: usr/lib/cca_stdll/cca_specific.c [hunk 4]
```diff
@@ -11708,10 +11725,22 @@ static CK_RV ccatok_wrap_key_rsa_pkcs(STDLL_TokData_t *tokdata,
     case CKK_AES:
         switch (mech->mechanism) {
         case CKM_RSA_PKCS:
+            if (keytype == sec_aes_cipher_key) {
+                TRACE_ERROR("CCA does not support to wrap with CKM_RSA_PKCS "
```
Comment (jschmidb) [inline usr/lib/cca_stdll/cca_specific.c:11859]: does not support wrapping with ... (?)

File: usr/lib/cca_stdll/ccatok.conf [hunk 5]
```diff
@@ -74,5 +74,10 @@ version cca-0
 # are generated or imported. Existing key objects that are backed by AES DATA
 # secure keys can still be used.
 #
+# Note: When CCA AES CIPHER keys are used, wrapping and unwrapping of such
+# AES CIPHER keys with mechanisms CKM_RSA_PKCS and CKM_RSA_PKCS_OAEP is not 
+# possible. Wrapping and unwrapping of CCA AES DATA keys is possible with those
+# mechnaisms.
```
Comment (jschmidb) [inline usr/lib/cca_stdll/ccatok.conf:80]: mechanisms

File: usr/lib/cca_stdll/cca_specific.c (hunk 4 above)
Comment (ifranzki) [reply usr/lib/cca_stdll/cca_specific.c:11859]: fixed

File: usr/lib/cca_stdll/ccatok.conf (hunk 5 above)
Comment (ifranzki) [reply usr/lib/cca_stdll/ccatok.conf:80]: fixed

File: testcases/misc_tests/cca_export_import_test.c [hunk 6]
```diff
@@ -748,7 +798,47 @@ static CK_RV cca_aes_data_export_import_tests(void)
...
+            goto error;
+        }
+        ilen = len = sizeof(data);
+        rc = funcs->C_Encrypt(session, data, ilen, encdata3, &len);
+        if (rc != CKR_OK) {
+            testcase_error("C_Encrypt rc=%s", p11_get_ckr(rc));
+            goto error;
+        }
+        if (ilen != len) {
+            testcase_fail("plain and encrypted data len does not match");
```
Comment (jschmidb) [inline testcases/misc_tests/cca_export_import_test.c:829]: do not match

Comment (ifranzki) [reply testcases/misc_tests/cca_export_import_test.c:829]: hm, well the 'does' is for the 'len', not for the 'plain and encrypted data', so IMHO 'does' is correct.

**Comment by ifranzki** [comment]
Finally, I found the reason for the travis failure with clang on s390x. It was clang warning about a set but unused variable. Unfortunately travis somehow was hiding the error message from clang, so that the make just failed, but without any (visible) compiler error message.

@jschmidb Please double check commit https://github.com/opencryptoki/opencryptoki/pull/814/commits/22bb92dc8c2146ae4b5030e39b776127b9668050 that fixes the set but unused variable in `pkey_aes_cmac()`. This variable was added with one of your LRG commits, but I guess the variable is obsolete.

**Comment by jschmidb** [comment]
Yes, you are right. Variable out_pos can be removed. The KMAC instruction writes its output into the parmblock, not into an output buffer. So in case of an LGR the whole operation can be retried. I think we could even remove in_pos and len, and let s390_kmac() operate repeatedly directly on message and message_len.

**Review by jschmidb — APPROVED** [review]

**Pull request merged**